python/lsst/cbp/coordinateConverter.py

```diff
@@ -0,0 +1,856 @@
+# This file is part of cbp.
+#
+# Developed for the LSST Data Management System.
+# This product includes software developed by the LSST Project
+# (https://www.lsst.org).
+# See the COPYRIGHT file at the top-level directory of this distribution
+# for details of code ownership.
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <https://www.gnu.org/licenses/>.
+
+__all__ = ["CoordinateConverter"]
+
+import math
+
+import numpy as np
+import scipy.optimize
+
+from . import coordUtils
+from lsst.geom import Extent2D, Point2D, SpherePoint, radians
+from lsst.afw.cameraGeom import PIXELS, FOCAL_PLANE, FIELD_ANGLE
+from .beamInfo import BeamInfo
+
+# Record errors from setFocalFieldAngle?
+_RecordErrors = False
+# list of errors from the root finder: error (abs value, radians), pupilPos, focalFieldAngle, beam
+_ErrorList = None
+
+
+def startRecordingErrors():
+    """Start recording setFocalFieldAngle errors and reset accumulators
+    """
+    global _RecordErrors, _ErrorList
+    _RecordErrors = True
+    _ErrorList = []
+
+
+def stopRecordingErrors():
+    """Stop recording errors"""
+    global _RecordErrors
+    _RecordErrors = False
+
+
+def getRecordedErrors():
+    """Return setFocalFieldAngle errors
+
+    Returns
+    -------
+    errorLists : `tuple` of three `list`
+
+        - rootFinderErrors: each element is a tuple:
+
+            - |error| in radians
+            - pupilPos
+            - focalFieldAngle
+            - beam name
+
+        - minimizerErrors: each element is a tuple
+          that is the same as root finder errors
+
+        - minimizerFailures: each element is a tuple:
+
+            - pupilPos
+            - focalFieldAngle
+            - beam name
+    """
+    return _ErrorList
+
+
+class CoordinateConverter:
+    """Coordinate conversions for the collimated beam projector (CBP)
+
+    This object supports the following tasks:
+
+    - Given a desired CBP "arrangement" (e.g. a particular beam
+      should fall on a particular spot on the pupil
+      and be focused to spot on a particular position of a detector),
+      compute the necessary telescope and CBP pointing
+      and the information about where each beam is directed.
+    - Given the current telescope and CBP pointing and a desired
```

Comment:
I agree that the term "beam arrangement" is a bit clumsy, but it is consistently used throughout. A beam arrangement consists of a desired position of one beam on the pupil and the position of the resulting spot (or, equivalently, the angle of incidence of the beam at the pupil), plus rotation of the camera to align the other spots properly on the focal plane.

To my ears this doesn't really sound like a "spot placement" but I could be convinced otherwise. If I fix it here I'll also fix in in many other places.

I actually prefer "beam configuration" to "beam arrangement", but the term "configuration" is too overloaded for me to be comfortable using it.

Reply:
Oh yes, I see. In that case the thing that had me confused is just that you're missing a "beam" in there, so just change `to the resulting CBP arrangement` to `to the resulting CBP beam arrangement` and then I think it reads fine, and stays consistent with everything else 🙂